Fix ftoi and disable VFPv3

JayFoxRox opened the pull request:
This fixes the printf bug (which is actually a bug in "dtoa")
The buggy code is:
```C
ds = (d2.d - 1.5) * 0.289529654602168 + 0.1760912590558 + i * 0.301029995663981;
k = (int) ds;
```
Example values were 0x3ff999999999999a for d2.d (double) and int i = -3.
When ds is -0.6, k would currently be 0x80000000 when 0 is expected (suggested by running dtoa on my x64).

Changing the 2 lines from `if (vdm.exponent >= 1023 - 1)` to just 1023 makes printf work fine (I think?) but I didn't really test it much.
You can find [similar code in qemu](https://github.com/qemu/qemu/blob/master/fpu/softfloat.c#L3126) which also compares using `if ( aExp < 0x3FF )` = `if (!(aExp >= 0x3FF))` = `if (!(aExp >= 1023))`.
There might be some shifting of the mantissa which makes this change incorrect so I'd prefer if someone could look over it.
However, the [single-precision code does it also like this PR suggests](https://github.com/citra-emu/citra/blob/master/src/core/arm/skyeye_common/vfp/vfpsingle.cpp#L587) already.

Fixes #1218 

---

Furthermore the interpreter contains some VFP3 instructions which cause #1483 . This is an entirely different problem but it was found by accident while searching for the other bug.
The idea for this change came from @MerryMage who said it was important that some instructions are ambigious. So I figured that VFP3 might "overwrite" some other instructions.
Information that [3DS only supports VFP2 came from @lioncash ](http://lioncash.github.io/ARMBook/vfp/README.html)
Is this correct or is there any reason for VFP3 code to exist?

Fixes #1483 

---

Tiny change but I'd be happy to split this into 2 seperate commits and give appropiate commit names. Suggestions?

lioncash commented:
Assigning myself to hardware test duty

JamePeng commented:
hmm,the ARM11(ARMv6) was not support the VFPv3, it implemented on most Cortex-A8 and A9 ARMv7 processors.
https://en.wikipedia.org/wiki/ARM_architecture#Floating-point_.28VFP.29

JayFoxRox commented on src/core/arm/skyeye_common/vfp/vfpdouble.cpp:
```diff
@@ -560,7 +560,7 @@ static u32 vfp_double_ftoui(ARMul_State* state, int sd, int unused, int dm, u32
     if (vdm.exponent >= 1023 + 32) {
```
Chances are this is supposed to be 1023 + 64 ?

JayFoxRox commented on src/core/arm/skyeye_common/vfp/vfpdouble.cpp:
```diff
@@ -644,7 +644,7 @@ static u32 vfp_double_ftosi(ARMul_State* state, int sd, int unused, int dm, u32
         if (vdm.sign)
```
(Slightly above where GitHub doesn't allow commenting)

Same: maybe should be 1023 + 64

JayFoxRox commented:
It's almost been a month. Progress update?
I need this for most HW tests (unless I want people to use compilers with softfp etc. or I change my HW tests to all int math).

lioncash commented:
Will make time for it tonight when I get home. Can you see if this issue still exists (can't do it myself since it's all at home). Curious if #1653 could have fixed it somehow, since it's possible a jump address was getting trashed.

JayFoxRox commented:
Nope, as I said before already: it's exactly that instruction + vfp3 bug.

The ftoi code looks fishy compared to the other similar code pieces (I linked to above) and that instruction produces absolutly ridiculous results (iirc: ftoi 0.3 ~~> (int)<giant value> instead of the expected 0(?))
(also see the dtoa code snippet above, I have actually followed the disassembly of it and traced it down to a handful of instructions - and the most promising candidate was ftoi where a quick comparison to other code pieces hinted what could be wrong)

The bug in Samurai Warriors (VFP3 issue) is also still present.

lioncash commented:
I'm somewhat skeptical this is a correct fix particularly because the Linux kernel emulates the soft-float routines with the same code, as seen [here](https://github.com/torvalds/linux/blob/master/arch/arm/vfp/vfpdouble.c#L528) and [here](https://github.com/torvalds/linux/blob/master/arch/arm/vfp/vfpdouble.c#L610). But anyway, I'll still test it when I get home.

Subv commented:
Reminder that this is still a thing

JayFoxRox commented:
I've written [a hardware test for this](https://github.com/JayFoxRox/3ds-tests/commit/a424220e82c8a7aa656877fa25b940003a5e73ab) now and could confirm my changes are better than master.
However, there were 2 issues:

1. The 3DS hangs after around 4000 randomly (some heuristics included) chosen doubles on some denormal [if not both, DN and FZ, are set in fpscr (click for doc)](http://infocenter.arm.com/help/index.jsp?topic=/com.arm.doc.ddi0360f/CHDFAFIA.html). The mode works fine in Citra and I can't see how this should hang the hardware. I might contact ARM about this because it looks like a hw issue to me.
2. With my fix only 3/4 rounding modes worked. The round-to-nearest mode (which was also broken in master) needed a special fix and I just looked at some numbers to figure out the current solution.

I have updated this PR with a fix for 2. and split the original commit into 2 commits, one for each feature.

I did not check the fpscr bits (inexact bits etc.) after running ftoi-instructions yet but I intend to do so. I also plan to test more numbers. I also hope @lioncash can help with 1. and I also hope he can do some extra tests due to how fishy this function seems to be.

JayFoxRox commented:
**Update:**

I've added NaN checking and denormal checking to my test now. Also I've run 100k tests per fpscr value (5 values), one run for ftoi32, ftoi64 (2 variants). => 5 * 2 * 100k tests.
I tried to run other modes but the hardware crash prevents me from testing without FZ or DN bit.
For the heuristics of input numbers you can [check this comment](https://github.com/JayFoxRox/3ds-tests/blob/cdb520a2ca0cc7651900ea250b31a931adb106dd/ftoi/source/main.c#L19)
Each test consists of 2 instructions: ftoiu + ftois, the fpscr is set before every instruction and read back afterwards. The output is one line per test, containing the inputer, result of the 2 ftoi and the 2 fpscr after execution.

The following is a comparison of output generated by @wwylele using [this version of the hw-test](https://github.com/JayFoxRox/3ds-tests/blob/cdb520a2ca0cc7651900ea250b31a931adb106dd/ftoi/source/main.c)

This is the number of lines output by diff for comparing wwyleles files against Citra `master` and this PR ("smdc" in the list) output.
A single line difference in the test output will cause 4 diff output lines.
Unfortunaly there is a bug in exception propagation in Citra (seperate PR coming soon). My test has 1/7 denormals => 57142.8571429 denormal lines which would cause the denormal flag in fpscr.
(As there are also totally random inputs in the test we get a couple more errors which are also denormals)

**However, with the exception propagation PR coming soon this will cause 0 errors.**
You can also see that a fpscr of 0x0300009F will produce 0 errors because the denormal flag is set anyway, so the missing exception propagation won't have any effect.

```
Result for Hardware (wwylele)/out-ftoi32-0x03000000.txt:
  master: 60356
  sdmc: 57340
Result for Hardware (wwylele)/out-ftoi32-0x0300009F.txt:
  master: 2716
  sdmc: 0
Result for Hardware (wwylele)/out-ftoi32-0x03400000.txt:
  master: 59420
  sdmc: 57340
Result for Hardware (wwylele)/out-ftoi32-0x03800000.txt:
  master: 74766
  sdmc: 57340
Result for Hardware (wwylele)/out-ftoi32-0x03C00000.txt:
  master: 59420
  sdmc: 57340
Result for Hardware (wwylele)/out-ftoi64-0x03000000.txt:
  master: 57850
  sdmc: 57164
Result for Hardware (wwylele)/out-ftoi64-0x0300009F.txt:
  master: 1356
  sdmc: 0
Result for Hardware (wwylele)/out-ftoi64-0x03400000.txt:
  master: 57850
  sdmc: 57164
Result for Hardware (wwylele)/out-ftoi64-0x03800000.txt:
  master: 72900
  sdmc: 57164
Result for Hardware (wwylele)/out-ftoi64-0x03C00000.txt:
  master: 57850
  sdmc: 57164
```

I didn't write this based on mathematical principles but just by observing the remaining errors and adding more code paths to work around issues.
The code quality is still pretty bad and I'm open to suggestions.

However, the hw-test shows that this is FAR better than this works correctly under lab-conditions and unless the tests are chosen poorly this confirms the functionality.
I've also tried to run a couple of games and they still work.

bunnei commented:
Code looks fine to me. Is this waiting on any additional testing or anything?
